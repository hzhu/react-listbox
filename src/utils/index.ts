import { SyntheticEvent } from "react";

export const KEY_CODES: Record<string, number> = {
  BACKSPACE: 8,
  TAB: 9,
  RETURN: 13,
  ESC: 27,
  SPACE: 32,
  PAGE_UP: 33,
  PAGE_DOWN: 34,
  END: 35,
  HOME: 36,
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40,
  DELETE: 46,
};

export type Handlers = (Function | undefined)[];

export const composeEventHandlers = (...handlers: Handlers) => (
  e: SyntheticEvent
) => {
  handlers.forEach((handler) => handler && handler(e));
};

export const focusElement = (
  element: HTMLElement | null,
  container: HTMLUListElement | null
) => {
  if (element === null || container === null) return;

  const scrollBottom = container.clientHeight + container.scrollTop;
  const elementBottom = element.offsetTop + element.offsetHeight;
  if (elementBottom > scrollBottom) {
    /* istanbul ignore next */
    container.scrollTop = elementBottom - container.clientHeight;
  } else if (element.offsetTop < container.scrollTop) {
    /* istanbul ignore next */
    container.scrollTop = element.offsetTop;
  }
};
