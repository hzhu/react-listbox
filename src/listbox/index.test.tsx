import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Listbox, ListboxOption } from "./";
import { KEY_CODES } from "../utils";

describe("Listbox", () => {
  test("renders a listbox (snapshot)", () => {
    const { getByRole } = render(
      <Listbox>
        <ListboxOption value="ford">Ford</ListboxOption>
        <ListboxOption value="tesla">Tesla</ListboxOption>
        <ListboxOption value="toyota">Toyota</ListboxOption>
      </Listbox>
    );
    const listbox = getByRole("listbox");

    expect(listbox).toMatchSnapshot();
  });
  describe("Uncontrolled", () => {
    test("calls the onChange & onSelect prop with option's value for keyboard selection", () => {
      const onChange = jest.fn();
      const onSelect = jest.fn();
      const { getByRole } = render(
        <Listbox onSelect={onSelect} onChange={onChange}>
          <ListboxOption value="ford">Ford</ListboxOption>
          <ListboxOption value="tesla">Tesla</ListboxOption>
          <ListboxOption value="toyota">Toyota</ListboxOption>
        </Listbox>
      );
      const listbox = getByRole("listbox");

      expect(onChange).toBeCalledTimes(0);

      fireEvent.focus(listbox);

      expect(onChange).toBeCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith("ford");

      fireEvent.keyDown(listbox, {
        keyCode: KEY_CODES.DOWN,
        which: KEY_CODES.DOWN,
      });

      expect(onChange).toHaveBeenCalledWith("tesla");
      expect(onChange).toBeCalledTimes(2);
      expect(onSelect).toBeCalledTimes(0);

      fireEvent.keyDown(listbox, {
        keyCode: KEY_CODES.DOWN,
        which: KEY_CODES.DOWN,
      });

      expect(onChange).toHaveBeenCalledWith("tesla");
      expect(onChange).toBeCalledTimes(3);
      expect(onSelect).toBeCalledTimes(0);

      fireEvent.keyDown(listbox, {
        keyCode: KEY_CODES.RETURN,
        which: KEY_CODES.RETURN,
      });

      expect(onChange).toBeCalledTimes(3);
      expect(onChange).toHaveBeenCalledWith("tesla");
      expect(onSelect).toBeCalledTimes(1);
      expect(onSelect).toHaveBeenCalledWith("toyota");
    });
    test("calls the onChange & onSelect prop with option's value for click selection", () => {
      const onChange = jest.fn();
      const onSelect = jest.fn();
      const { getByText } = render(
        <Listbox onSelect={onSelect} onChange={onChange}>
          <ListboxOption value="tesla">Tesla</ListboxOption>
        </Listbox>
      );
      const tesla = getByText("Tesla");

      fireEvent.click(tesla);

      expect(onChange).toBeCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith("tesla");
      expect(onSelect).toBeCalledTimes(1);
      expect(onSelect).toHaveBeenCalledWith("tesla");
    });

    test("able to composes and call ListboxOption's onClick handler", () => {
      const onClick = jest.fn();
      const { getByText } = render(
        <Listbox>
          <ListboxOption value="tesla" onClick={onClick}>
            Tesla
          </ListboxOption>
        </Listbox>
      );
      const tesla = getByText("Tesla");

      fireEvent.click(tesla);

      expect(onClick).toBeCalledTimes(1);
    });
  });
});
