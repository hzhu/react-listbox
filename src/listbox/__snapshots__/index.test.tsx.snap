// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listbox renders a listbox (snapshot) 1`] = `
<ul
  role="listbox"
  tabindex="0"
>
  <li
    id="option--ford--1"
    role="option"
  >
    Ford
  </li>
  <li
    id="option--tesla--2"
    role="option"
  >
    Tesla
  </li>
  <li
    id="option--toyota--3"
    role="option"
  >
    Toyota
  </li>
</ul>
`;
